### ISSUE TITLE
permalink / output location / url separation?

### ISSUE THREAD

**Issue by rendall** [11ty/eleventy#342]
Hello!

TL;DR need a way to modify or specify the url for a file, distinct from its output directory. Or, if it exists, I am willing to update the documents.

I am also willing to code this up if it does not  exist. 

The `/blog` subdirectory in my site's file structure is the root directory of a `blog.example.com` subdomain, and whatever is above that should be in top level domain. However, all of the collections in the `/blog` subdirectory have urls with `/blog` prepended (e.g. `blog.example.com/blog/foo` rather than `blog.example.com/foo`).

There does not appear to be an easy way to specify a url that is different from output directory: the assumption appears to be that, say, outputting to a `/blog` subdirectory will necessarily mean that `/blog` will be in the path of all urls.  I thought that `permalink` in the front matter might help, i.e. the files with the template 'post.njk' in `/blog/posts` could be linked to just `/posts/*` by changing their `permalink` value, and indeed it changes the url, but also it changes the output directory to `/posts` (as a sibling to the `/blog` directory, not a subdirectory of it).

I attempted to alter the `url` values of the posts collection items in the `.eleventy.js` file, and indeed the `addCollection` callback returns objects with `{...post, url:"/posts/foo/"}` correct values, but it's no bueno. Wherever `post.url` is used in the template uses the directory structure (i.e. prepends `/blog`).

I'm certain I am missing something, and if so I will happily submit a PR to update the documents. But if not, I recommend having a clear way to distinguish the url used to link to the item from its output directory. 

Thanks so much!

**Comment by zachleat** [comment]
Hmm, I’m not completely sure what you mean here. Can you take me through the following in your example?

1. Input Directory
2. Output Directory
3. Template Filename
4. Desired template output location

**Comment by zachleat** [comment]
Just guessing based on what I think I’m understanding, I’m recommend checking out these two things:
* pathPrefix configuration option: https://www.11ty.io/docs/config/#deploy-to-a-subdirectory-with-a-path-prefix)
* permalink documentation on ignoring the output directory: https://www.11ty.io/docs/permalinks/#ignore-the-output-directory

**Comment by rendall** [comment]
My apologies for being unclear, and thank you for the links!  I will check them out and also find a clearer expression of what I am trying to do!

**Comment by rendall** [comment]
Thank you so, so much for your help, Zach. Truly appreciated. If you need any help on this whatsoever, please ask.

To directly answer your question, my single project file structure maps to two distinct sites: `example.com` and `blog.example.com` in this way:

1. Input Directory: `project\_src`

2. Output Directory: `project\`

3. Template Filename: (some examples: `index.md`, `cv.md`, `project\_src\blog\posts\hello-world.md`)

4. Desired template output location
`project\**` => `example.com/**` except:
`project\blog\**` => `blog.example.com/**`

If that's not clear, here is a hopefully not too convoluted description of what I'm struggling with:

(For clarity, I'll distinguish local project directories by backslashes `\`  and URLs with forwardslashes `/`. `**` just means 'any files and subdirectories'.)

Eleventy source files sit inside of `project\_src\**` (which outputs to `project\**`)  and that includes `project\_src\blog\**` which as expected outputs to `project\blog\**`.  (So, nearly the opposite of 11ty's default expectation of `project\**` to output to `project\_site\**`).

Moreover, the URLs map in this way:
`project\**` maps to `example.com/**`  and
`project\blog\**` maps to `blog.example.com/**`

Files coming from `project\_src\blog` have a different needs for mapping URLs. By default, eleventy expects the URLs of files in `project\_src\**` to be `example.com/**`, and that includes files in `project\_src\blog\**` which would ordinarily map to `example.com/blog/**`. 

However, since `project\blog\**` actually maps to `blog.example.com/**` I need to change the URL mapping for any file in `project\_src\blog\` while keeping the default mapping for files not in the `blog` subdirectory.

So, for specific examples, I would like a 'home' link in `project\cv\index.html` to map to the URL `/` and a 'cv' link in `project\index.html` to map to the URL `/cv` (i.e. `<a href='/cv'>cv</a>`. So far, so good. Separately, I would like a 'home' link in `project\blog\post\hello-world\index.html` to map to `/` (*not* `/blog`) and a 'hello world' link in `project\blog\index.html` to link to `/posts/hello-world` (*not* `/blog/posts/hello-world`).

**Comment by rendall** [comment]
From your links:

* the [deploy to a subdirectory](https://www.11ty.io/docs/config/#deploy-to-a-subdirectory-with-a-path-prefix) page seems to be how to change the URL for the entire site. It is related, but I am struggling with how to do this for a single subdirectory and not the entire site. If I can apply this to a single subdirectory, that would be amazing.
* The [ignore output directory](https://www.11ty.io/docs/permalinks/#ignore-the-output-directory) page seems to be about how to change the output directory for a specific file, which is not exactly my struggle.

However, I wonder if having two .eleventy.js config files will do the trick?  I will have one in `project\_src\` that takes care of `project\_src\**` and another one that sits inside of `project\_src\blog\` that takes care of `project\_src\blog\**`.

I bet that will do it!

**Comment by rendall** [comment]
OK! I now have two sites using the non-conventional project structure above with a single build process.  I'll outline the method below. The TL;DR general trick is to call eleventy twice, once in each subdirectory.

First and foremost, though: thank you, @zachleat + contributors, for such a wonderful static generator and open-source project. There are not enough thank-yous that I can possibly send. Moreover, if anything I said, or will say, dismays, confuses, confounds or even annoys you, please ignore it and only remember that you have my gratitude, full stop. If you need help, just ask and I will take care of it.

I'll outline the solution, below, in case anyone needs it.

First, rather than the expected, default configuration of `input=.` and `output=./_site/`, I have `input=./_src` and `output=.`, which is exactly the opposite direction of outputting the project directory into a subdirectory.  To handle this, cd into the source subdirectory and output to the parent directory, i.e. `cd _src && eleventy --input=. --output=../`

Second, I have a subdirectory `./_src/blog` that needs to output to `./blog` but the URLs need to be prepended with `/` and not `/blog/` (as well as [other the other blog-specific configurations and plugins](https://github.com/11ty/eleventy-base-blog)). To handle that case, 
* add the line `blog/` to `./_src/.eleventyignore`
* make sure that `blog/` has its own `_includes`, `_data`, `_11ty` directories and so forth
* add a `./_src/blog/.eleventy.js` config file with `input: "."`, `pathprefix: "/", and `output: "../../blog/"`
* the command is `cd _src/blog && eleventy`

**Comment by zachleat** [comment]
Thank you so much for your words and sharing your solution!
